internal/csot/csot.go

```diff
@@ -11,26 +11,111 @@ import (
 	"time"
 )
 
-type timeoutKey struct{}
+type withoutMaxTime struct{}
 
-// MakeTimeoutContext returns a new context with Client-Side Operation Timeout (CSOT) feature-gated behavior
-// and a Timeout set to the passed in Duration. Setting a Timeout on a single operation is not supported in
-// public API.
-//
-// TODO(GODRIVER-2348) We may be able to remove this function once CSOT feature-gated behavior becomes the
-// TODO default behavior.
-func MakeTimeoutContext(ctx context.Context, to time.Duration) (context.Context, context.CancelFunc) {
-	// Only use the passed in Duration as a timeout on the Context if it
-	// is non-zero.
-	cancelFunc := func() {}
-	if to != 0 {
-		ctx, cancelFunc = context.WithTimeout(ctx, to)
+// WithoutMaxTime returns a new context with a "withoutMaxTime" value that
+// is used to inform operation construction to not include "maxTimeMS" in a wire
+// message, regardless of a context deadline. This is specifically used for
+// non-awaitable hello commands.
+func WithoutMaxTime(ctx context.Context) context.Context {
+	return context.WithValue(ctx, withoutMaxTime{}, true)
+}
+
+// IsWithoutMaxTime checks if the provided context has been assigned the
+// "withoutMaxTime" value.
+func IsWithoutMaxTime(ctx context.Context) bool {
+	return ctx.Value(withoutMaxTime{}) != nil
+}
+
+type clientLevel struct{}
+
+func AsClientLevel(parent context.Context) context.Context {
+	return context.WithValue(parent, clientLevel{}, true)
+}
+
+func IsClientLevel(ctx context.Context) bool {
+	val := ctx.Value(clientLevel{})
+	if val == nil {
+		return false
 	}
-	return context.WithValue(ctx, timeoutKey{}, true), cancelFunc
+
+	return val.(bool)
 }
 
+// IsTimeoutContext checks if the provided context has been assigned a deadline
+// or has unlimited retries.
 func IsTimeoutContext(ctx context.Context) bool {
-	return ctx.Value(timeoutKey{}) != nil
+	_, ok := ctx.Deadline()
+
+	return ok || IsClientLevel(ctx)
+}
+
+// WithTimeout will set the given timeout on the context, if no deadline has
+// already been set.
+//
+// This function assumes that the timeout field is static, given that the
+// timeout should be sourced from the client. Therefore, once a timeout function
+// parameter has  been applied to the context, it will remain for the lifetime
+// of the context.
+func WithTimeout(parent context.Context, timeout *time.Duration) (context.Context, context.CancelFunc) {
+	cancel := func() {}
+
+	// In the following conditions, do nothing:
+	//	1. The parent already has a deadline
+	//	2. The parent does not have a deadline, but a client-level timeout has
+	//		 been applied.
+	//	3. The parent does not have a deadline, there is not client-level timeout,
+	//     and the timeout parameter DNE.
+	if IsTimeoutContext(parent) || timeout == nil {
```

Comment:
Optional: Reverse the condition here to optimize for the case that no client-level timeout is set, which is currently the most common case.
```suggestion
	if timeout == nil || IsTimeoutContext(parent) {
```